### PR TITLE
Event Hubs Processor - Add Sample Using the EventProcessorClient with an ASP.NET hosted service

### PR DESCRIPTION
**Pull request by davetrainer**
Addresses [26299](https://github.com/Azure/azure-sdk-for-net/issues/26299)

An example project which demonstrates how to use the EventProcessorClient with an ASP.NET hosted service to accomplish background task processing.

### CODE REVIEW

**Comment by davetrainer** [comment]
@microsoft-github-policy-service agree

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:4]:
```suggestion
    <RequiredTargetFrameworks>net8.0;net6.0</RequiredTargetFrameworks>
```
We may as well switch to `net8.0`, as it's now GA.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:1]: If you'd be so kind, let's shift to a folder structure that allows for the sample and its associated tests to be included here.  For example, the [KeyVault proxy sample](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/keyvault/samples/keyvaultproxy) uses the structure that I'm thinking of - though we'll want to keep the README at the root with its current content.

```text
sdk/eventhub/Azure.Messaging.EventHubs.Processor
|
| -- samples
        | -- AspNetHostedService
                | -- Solution file 
                | -- README.md (currently sample_AspNetCore.md)
                | -- Directory.Build.targets  (version overrides when outside the repository; optional - doing it inline in each project is fine)
                | -- src
                        | -- sample project and assets, less the markdown file
                | -- tests
                       | -- test project (new)
                       | -- HostedServiceTests.cs (hoisted from main tests project)
```

I leave it to your discretion whether you'd like to include an explicit gitignore, editorconfig or any of the other "I downloaded this project in a zipfile" assets that the KeyVault sample has.  Likewise, I'm good if you'd prefer to keep the conditional version assignment locally in each project file or extract to a `Directory.Build.targets` file.

**Review by jsquire — COMMENTED** [review]
Hi @davetrainer.  Thank you very much for your contribution and your interest in improving the Azure developer experience.   The sample looks great!   

I've left some feedback asking to align a few things closer to repository conventions, if you'd be so kind.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj]: Directory.Build.targets added

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Azure.Messaging.EventHubs.Processor.Tests.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Azure.Messaging.EventHubs.Processor.Tests.csproj]: Done

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj]: @jsquire I've left this as net7.0 as I had issues with the SDK. I believe globals.json needs updating first? https://github.com/Azure/azure-sdk-for-net/blob/main/global.json

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj:1]: Done. Left out gitignore and editorconfig

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs]: Done

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/SampleProgram.cs]: Done

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/sample_AspNetCore.md
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/sample_AspNetCore.md]: Left as is due to net8 SDK issue

**Comment by davetrainer** [comment]
@jsquire Thanks for the detailed and thorough review! I've made the necessary changes. I was unable to get the solution building when targeting net8.0, due to an SDK mismatch. I wasn't bale to resolve it. If there's a way around this, I can incorporate it if needed.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Azure.Messaging.EventHubs.Samples.Processor.HostedService.csproj]: That's my bad.  I'm getting ahead of myself and forgetting that we haven't updated the SDK version yet.  Leaving it as-is makes sense.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/README.md
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/README.md:84]: This URL will likely continue to cause analysis to fail because it is correctly an explicit reference to `main`  but the path doesn't yet exist.   Once we get to the point where it's our only remaining failure, we can decide whether to force bypass of the failure or comment it out and bring it back after this merges.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/AspNetHostedService.sln
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/AspNetHostedService.sln:1]: Not sure if the path was already updated and not showing in the GH UI, but can we please confirm that we've adjusted case to `AspNetHostedService`?

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets:8]: We don't want the condition here; this will need to link back up to the root targets file in order to keep the needed integration with the repo engineering process and systems.   We'll need to adjust the path to climb up a few more subdirectories.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/Directory.Build.targets:3]: We'll want to be sure to do this after the import, so that we're covered in case the definition of `IsSamplesProject` moves around to the repo targets instead of the props where it lives today.   Unlikely that it would happen, but there's no downside to flipping the order here.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:12]:
```suggestion
# Using the `EventProcessorClient` with an ASP.NET hosted service

```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:20]:
```suggestion
### Prerequisites

```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/README.md:14]:
```suggestion
This is an example of how to create and configure the `EventProcessorClient` in an ASP.NET context as an `IHostedService` implementation.  ASP.NET hosted services can be used to implement background task processing, making them a good choice if your application needs to continually process events from Azure Event Hubs.  For more information, please see the [hosted services documentation](https://learn.microsoft.com/aspnet/core/fundamentals/host/hosted-services).
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj:4]: This will fail outside the Azure SDK repository.   We'll need to do a conditional definition to define these if we're not in the repo.   Since we've got a targets file, that'd be the place.

**Review by jsquire — COMMENTED** [review]
Thanks, @davetrainer.   We're looking good!  There's a few minor nits and some adjustments  for accomodating the dual nature of being inside and outside of the repo that I've asked for.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj:4]: @jsquire I'm not fully understanding the ask here. Can you give me an example of what you're looking for? I've looked for other examples in sample projects but can't find anything.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/AspNetHostedService.sln
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/AspNetHostedService.sln:1]: @jsquire Folder path is now correctly cased

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/src/SampleProgram.cs
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/src/SampleProgram.cs]: @jsquire The SDK build system is enforcing XML comments, so I have to add something. I've updated where needed.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj]: @jsquire Understood. I've added `Directory.Build.props` and updated `Directory.Build.targets`. I'm forced to only target net6 and net7 in `Directory.Build.props`, as it's not possible to build against net462 for the aspnet app (net462 get's passed down by `Directory.Build.Common.props` , but I have to override ride it).  If there's a way to work around this, please let me know.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props:3]:
```suggestion
  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\.., Directory.Build.props))\Directory.Build.props"
```

If I'm reading correctly, I think this climbs 4 directories up, which puts us in the `/sdk` directory.  Since we want to be in `/sdk/eventhub`  I believe we need to trim one level.

Please double check me.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props:4]:
```suggestion
			Condition="'$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\.., Directory.Build.props))' != ''" />
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/tests/Azure.Messaging.EventHubs.Processor.HostedService.Samples.Tests.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/tests/Azure.Messaging.EventHubs.Processor.HostedService.Samples.Tests.csproj:3]: Should we consider only including `net462` when building from the Azure SDK repo?  It really exists just to service our common text matrix.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/Azure.Messaging.EventHubs.Processor.HostedService.Samples.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/Azure.Messaging.EventHubs.Processor.HostedService.Samples.csproj:3]: Can we please rename to `Azure.Messaging.EventHubs.Processor.Samples.HostedService`?   

The Event Hubs overrides allow for this pattern to also set the `IsSamplesProject` flag that we're using for detecting the build location.  _([src](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Directory.Build.props#L10))_

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:37]:
```suggestion
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:36]: Please add a description here.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:52]: This is true of the event handlers assigned to the processor, but I think we want to throw here if the processor has issues starting, no?   For example, if the connection string is incorrect, we know that the processor will never recover.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:161]:
```suggestion
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/EventProcessorClientService.cs:160]: Please add text for this one.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:30]:
```suggestion
            return Task.CompletedTask;
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:26]:
```suggestion
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleApplicationProcessor.cs:43]:
```suggestion
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs:16]: We may want to remove the args parameter, since we're not actually accepting anything.   Otherwise, please add text to the doc comment.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/src/SampleProgram.cs:48]:
```suggestion
        private static EventProcessorClient CreateEventProcessorClient(IConfiguration configuration)
```
nit: "Get" could be interpreted as grabbing an existing instance from the DI container or other cache.

**Review by jsquire — COMMENTED** [review]
Apologies for the delay, @davetrainer.   This is looking great!   I started digging into the lower-level details this time out, and there are just a few small nits that I left feedback on.   I think we're just about ready!

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/aspnethostedservice/tests/Azure.Messaging.EventHubs.Processor.Samples.HostedService.Tests.csproj]: The only way to work around it would be to screen it out unless `IsSamplesProject` is set;  the `net462` target is part of our required test matrix in the repo.   I think the approach that you've got for just hiding the tests is the best we can do.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/Directory.Build.props:3]: @jsquire  I can confirmed this is the correct change.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/tests/Azure.Messaging.EventHubs.Processor.HostedService.Samples.Tests.csproj
Comment (davetrainer) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/AspNetHostedService/tests/Azure.Messaging.EventHubs.Processor.HostedService.Samples.Tests.csproj:3]: @jsquire Agreed. I've added this to the props file.

**Comment by davetrainer** [comment]
> Apologies for the delay, @davetrainer. This is looking great! I started digging into the lower-level details this time out, and there are just a few small nits that I left feedback on. I think we're just about ready!

Hi @jsquire Happy new year! I have addressed all comments. Could you please take another look? It would be great if I could get this merged.

File: eng/Packages.Data.props
Comment (jsquire) [inline eng/Packages.Data.props:292]: Any reason that we shouldn't push these to the latest?   I believe the 8.0 line should work just fine with `net6.0` and `net7.0` targets.  Not particularly important nor blocking.   I can always bump them in a follow-up.

**Comment by jsquire** [comment]
> Hi @jsquire Happy new year! I have addressed all comments. Could you please take another look? It would be great if I could get this merged.
> 
> I've tested inside and out of the repo structure, the app works as expected in both.

Thanks, @davetrainer!   Your efforts and contribution are very much appreciated.   This looks great.   I'm kicking the (unrelated) failed bits in CI to see if I can get them to pass on re-run.   If not, I'll just override the failure and get things merged!

**Review by jsquire — APPROVED** [review]

File: eng/Packages.Data.props
Comment (davetrainer) [reply eng/Packages.Data.props:292]: @jsquire 7.0 would've been what I got when VS added the nuget package reference - I assume as it's targets net 7, it picked that one. I'm happy to bump and retest if you want me to?

Comment (jsquire) [reply eng/Packages.Data.props:292]: @davetrainer:  I appreciate your willingness, but no need.   I'll need to do a follow-up to uncomment the new sample link once it's live.   I'll play with the package references when I do that one.   I'm just waiting on one other thing to close internally and then I'll get this merged in.

Comment (jsquire) [inline eng/Packages.Data.props:292]:
```suggestion
	 <PackageReference Update="Microsoft.Extensions.Configuration.Abstractions" Version="8.0.0" />
```

File: eng/Packages.Data.props
Comment (jsquire) [inline eng/Packages.Data.props:296]:
```suggestion
    <PackageReference Update="Microsoft.Extensions.Hosting" Version="8.0.0" />	
```

**Pull request merged**
